pop_tools/zonal_average.py

```diff
@@ -0,0 +1,247 @@
+            self.grid = get_grid(grid_name)
+
+        elif grid is not None:
+            self.grid = grid
+
+        else:
+            raise ValueError('Failed to input grid name or grid dataset')
+
+        # Set the dx/dy parameters for generating the grid
+        self.dx = dx
+        self.dy = dy
+
+        # Set the regridding method
+        self.regrid_method = regrid_method
+
+        # Set the grid generation method
+        self.method_gen_grid = method_gen_grid
+
+        # If the user does not input a mask, use default mask
+        if not mask:
+            self.mask = self.grid['REGION_MASK']
+
+            if 'region_name' not in self.grid.variables:
+                self.mask_labels = np.unique(self.mask.values)
+
+            else:
+                self.mask_labels = self.grid['region_name']
+
+        else:
+            self.mask = mask
+
+    # Setup method for regridding a dataarray
+    def _regrid_dataarray(self, da_in, regrid_mask=False, regrid_method=None):
+
+        src_grid = _convert_to_xesmf(da_in, self.grid)
+        dst_grid = _generate_dest_grid(self.dy, self.dx, self.method_gen_grid)
+
+        # If the user does not specify a regridding method, use default conservative
+        if regrid_method is None:
+            regridder = _generate_weights(src_grid, dst_grid, self.regrid_method)
+
+        else:
+            regridder = _generate_weights(src_grid, dst_grid, regrid_method)
+
+        # Regrid the input data array, assigning the original attributes
+        da_out = regridder(src_grid[da_in.name])
+        da_out.attrs = da_in.attrs
+
+        return da_out
+
+    def regrid(self, obj, **kwargs):
+        """generic interface for regridding DataArray or Dataset"""
+        if isinstance(obj, xr.Dataset):
+            ds_list = []
+            for var in obj.variables:
+
+                # Make sure the variable has the correct dimensions, is not a coordinate, and is not a velocity
+                if (
+                    ('nlat' and 'nlon' in obj[var].dims)
+                    and (var not in list(obj.coords))
+                    and ('velocity' not in obj[var].long_name.lower())
+                ):
+                    ds_list.append(obj[var])
+            return xr.merge(ds_list).map(self._regrid_dataarray, keep_attrs=True, **kwargs)
+        elif isinstance(obj, xr.DataArray):
+            return self._regrid_dataarray(obj, **kwargs)
+        raise TypeError('input data must be xarray DataArray or xarray Dataset!')
+
+    def zonal_average(self, obj, vertical_average=False, **kwargs):
+
+        data = self.regrid(obj, **kwargs)
+        mask = self.regrid(self.mask, regrid_method='nearest_s2d', **kwargs)
+
+        # Store the various datasets seperated by basin in this list
+        ds_list = []
+        for region in tqdm(np.unique(mask)):
+
+            if region != 0:
+                ds_list.append(data.where(mask == region).groupby('lat').mean())
```

Comment:
Why groupby? Just a simple `.mean("lat")`.

instead of this loop, I would recommend taking a 2D mask variable and expanding out to 3D with a new `region` dimension (so that each slice along region has 1's only for the appropriate region) and then multiply. That way you're doing `where` on a small array rather than a giant one. It also avoids the `concat` down below

Reply:
Matt and were discussing, making use of the `pop_tools.region_mask_3d` tool would be helpful with this, replacing "0" with `nan` values... 
```
# Read in the 3D mask
mask3d = pop_tools.region_mask_3d(grid_name, mask_name='default')

# 
data_masked = mask3d.where(mask3d > 0) * data[list_relavant_variables]
```

where the dimensions of data_masked would be `('region', 'nlat', 'nlon', 'time', 'z_t')`